MatchServer/MMatchRuleGunGame.cpp

```diff
@@ -0,0 +1,134 @@
+#include "stdafx.h"
+#include "MMatchRuleGunGame.h"
+#include "MBlobArray.h"
+
+MMatchRuleGunGame::MMatchRuleGunGame(MMatchStage* pStage) : MMatchRule(pStage)
+{
+}
+
+MMatchRuleGunGame::~MMatchRuleGunGame()
+{
+}
+
+void MMatchRuleGunGame::OnBegin()
+{
+	if (m_pGGSet.size() > 0) {
+		for (auto itor = m_pGGSet.begin(); itor != m_pGGSet.end(); ++itor)
+		{
+			delete *itor;
+		}
+		m_pGGSet.clear();
+	}
+
+	int randSet = rand() % MGetGunGame()->GetSetCount();
+	m_pGGSet = MGetGunGame()->GetGGSet(randSet);
+}
+
+void MMatchRuleGunGame::OnRoundBegin()
+{
+
+}
+
+void MMatchRuleGunGame::OnEnd()
+{
+	for (auto itor = m_pGGSet.begin(); itor != m_pGGSet.end(); ++itor)
+	{
+		delete *itor;
```

Comment:
`GunGame::GetGGSet` returns a list of *copies* of the GGSet pointers. Therefore, `MMatchRuleGunGame` does not own them and shouldn't try to delete them, otherwise the original list would be full of dangling pointers.